analyzer/pom-analyzer/src/main/java/org/jboss/shrinkwrap/resolver/impl/maven/MavenResolvedArtifactImpl.java

```diff
@@ -0,0 +1,272 @@
+
+    private final File file;
+
+    private MavenResolvedArtifactImpl(MavenCoordinate mavenCoordinate, String resolvedVersion, boolean snapshotVersion,
+            String extension, File file, ScopeType scopeType, MavenArtifactInfo[] dependencies) {
+        super(mavenCoordinate, resolvedVersion, snapshotVersion, extension, scopeType, dependencies, false);
+        this.file = file;
+    }
+
+    private MavenResolvedArtifactImpl(final Artifact artifact, final ScopeType scopeType,
+            final List<DependencyNode> children, boolean optional) {
+        super(artifact, scopeType, children, optional);
+        this.file = artifactToFile(artifact);
+    }
+
+    
+    // <new> (1/3)
+    public static Set<ResolutionResult> artifactRepositories = null;
+    // </new>
+    
+    /**
+     * Creates MavenResolvedArtifact based on ArtifactResult.
+     *
+     * @param artifactResult
+     * @return
+     */
+    static MavenResolvedArtifact fromArtifactResult(final ArtifactResult artifactResult) {
+        final Artifact artifact = artifactResult.getArtifact();
+        final DependencyNode root = artifactResult.getRequest().getDependencyNode();
+
+        // <new> (2/3)
+        if(artifactRepositories != null) {
+        	String coordinate = artifact.toString();
+        	String url = getUrl(artifactResult.getRepository());
+        	File file = artifact.getFile();
+			artifactRepositories.add(new ResolutionResult(coordinate, url, file));
+        }
+        // </new>
+        
+        // SHRINKRES-143 lets ignore invalid scope
+        ScopeType scopeType = ScopeType.RUNTIME;
+        try {
+            scopeType = ScopeType.fromScopeType(root.getDependency().getScope());
+        } catch (IllegalArgumentException e) {
+            // let scope be RUNTIME
+            log.log(Level.WARNING, "Invalid scope {0} of retrieved dependency {1} will be replaced by <scope>runtime</scope>",
+                    new Object[] { root.getDependency().getScope(), root.getDependency().getArtifact() });
+        }
+
+        final List<DependencyNode> children = root.getChildren();
+        final boolean optional = root.getDependency().isOptional();
+        return new MavenResolvedArtifactImpl(artifact, scopeType, children, optional);
+    }
+
+    // <new> (3/3)
+    private static String getUrl(ArtifactRepository repository) {
+		if(repository instanceof RemoteRepository) {
+			return ((RemoteRepository)repository).getUrl();
+		} else {
+			return String.format("n/a (%s)", repository.getClass().getSimpleName());
+		}
+	}
+    // </new>
+    
+    @Override
+    public <RETURNTYPE> RETURNTYPE as(Class<RETURNTYPE> returnType) {
+        if (returnType == null) {
+            throw new IllegalArgumentException("Type must be specified.");
+        }
+
+        final FormatProcessor<? super MavenResolvedArtifact, RETURNTYPE> processor = FormatProcessors.find(
```

Comment:
My IDE flags a compile error at this line:
"no instance(s) of type variable(s) RESOLVEDTYPE exist so that Object conforms to ResolvedArtifact<RESOLVEDTYPE>"

A cast is needed to resolve it:
```
final FormatProcessor<? super MavenResolvedArtifact, RETURNTYPE> processor = (FormatProcessor<? super MavenResolvedArtifact, RETURNTYPE>) FormatProcessors.find(
    MavenResolvedArtifact.class, returnType);
```